Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+         "<MDocument><MChemicalStruct><reaction>";
+
+  out << "<reactantList>";
+  for (auto &reactant : reactants) {
+    out << reactant->toString();
+  }
+  out << "</reactantList>";
+  out << "<agentList>";
+  for (auto &agent : agents) {
+    out << agent->toString();
+  }
+  out << "</agentList>";
+  out << "<productList>";
+  for (auto &product : products) {
+    out << product->toString();
+  }
+  out << "</productList>";
+
+  out << arrow.toString();
+
+  out << "</reaction></MChemicalStruct>";
+
+  for (auto &pluse : pluses) {
+    out << pluse->toString();
+  }
+  for (auto &condition : conditions) {
+    out << condition->toString();
+  }
+
+  out << "</MDocument></cml>";
+  return out.str();
+}
+
+ptree MarvinReaction::toPtree() const {
+  ptree out;
+
+  out.put("cml.<xmlattr>.xmlns", "http://www.chemaxon.com");
+  out.put("cml.<xmlattr>.xmlns:xsi",
+          "http://www.w3.org/2001/XMLSchema-instance");
+  out.put(
+      "cml.<xmlattr>.xsi:schemaLocation",
+      "http://www.chemaxon.com http://www.chemaxon.com/marvin/schema/mrvSchema_20_20_0.xsd");
+
+  for (auto &reactant : reactants) {
+    out.add_child(
+        "cml.MDocument.MChemicalStruct.reaction.reactantList.molecule",
+        reactant->toPtree());
+  }
+
+  for (auto &agent : agents) {
+    out.add_child("cml.MDocument.MChemicalStruct.reaction.agentList.molecule",
+                  agent->toPtree());
+  }
+
+  for (auto &product : products) {
+    out.add_child("cml.MDocument.MChemicalStruct.reaction.productList.molecule",
+                  product->toPtree());
+  }
+
+  out.put_child("cml.MDocument.MChemicalStruct.reaction.arrow",
+                arrow.toPtree());
+
+  for (auto &pluse : pluses) {
+    out.add_child("cml.MDocument.MReactionSign", pluse->toPtree());
+  }
+
+  for (auto &condition : conditions) {
+    out.add_child("cml.MDocument.MTextBox", condition->toPtree());
+  }
+
+  return out;
+}
+
+MarvinRectangle::MarvinRectangle() {
+  upperLeft.x = 0.0;
+  upperLeft.y = 0.0;
+  lowerRight.x = 0.0;
+  lowerRight.y = 0.0;
+  centerIsStale = true;
+}
+
+MarvinRectangle::MarvinRectangle(double left, double right, double top,
+                                 double bottom) {
+  upperLeft.x = left;
+  upperLeft.y = top;
+  lowerRight.x = right;
+  lowerRight.y = bottom;
+  centerIsStale = true;
+}
+
+MarvinRectangle::MarvinRectangle(const RDGeom::Point3D &upperLeftInit,
+                                 const RDGeom::Point3D &lowerRightInit) {
+  upperLeft = upperLeftInit;
+  lowerRight = lowerRightInit;
+  centerIsStale = true;
+}
+
+MarvinRectangle::MarvinRectangle(const std::vector<MarvinAtom *> atoms) {
+  centerIsStale = true;
+
+  if (atoms.size() == 0) {
+    return;
+  }
+  upperLeft.x = DBL_MAX;
+  upperLeft.y = -DBL_MAX;
+  lowerRight.x = -DBL_MAX;
+  lowerRight.y = DBL_MAX;
+
+  for (auto atom : atoms) {
+    if (atom->x2 < upperLeft.x) {
```

Comment:
these kinds of things are simpler if you use std::min and std::max:
```
upperLeft.x = std::min(upperLeft.x,atom->x2);
```

Reply:
It could be less efficient  with min and max, but code is cleaner.  I changed it